Support base-4.12

I can't update `ekg-carbon` until `ekg-core` builds with `base-4.12`.

base-4.12 seems to have added additional fields which need to be taken care of in order to be compatible w/ base-4.12:

```
System/Metrics.hs:479:17: warning: [-Wmissing-fields]
    • Fields of ‘Stats.RTSStats’ not initialised: init_cpu_ns,
                                                  init_elapsed_ns
    • In the expression:
        Stats.RTSStats
          {gcs = 0, major_gcs = 0, allocated_bytes = 0, max_live_bytes = 0,
           max_large_objects_bytes = 0, max_compact_bytes = 0,
           max_slop_bytes = 0, max_mem_in_use_bytes = 0,
           cumulative_live_bytes = 0, copied_bytes = 0, par_copied_bytes = 0,
           cumulative_par_max_copied_bytes = 0,
           cumulative_par_balanced_copied_bytes = 0, mutator_cpu_ns = 0,
           mutator_elapsed_ns = 0, gc_cpu_ns = 0, gc_elapsed_ns = 0,
           cpu_ns = 0, elapsed_ns = 0, gc = emptyGCDetails}
      In an equation for ‘emptyRTSStats’:
          emptyRTSStats
            = Stats.RTSStats
                {gcs = 0, major_gcs = 0, allocated_bytes = 0, max_live_bytes = 0,
                 max_large_objects_bytes = 0, max_compact_bytes = 0,
                 max_slop_bytes = 0, max_mem_in_use_bytes = 0,
                 cumulative_live_bytes = 0, copied_bytes = 0, par_copied_bytes = 0,
                 cumulative_par_max_copied_bytes = 0,
                 cumulative_par_balanced_copied_bytes = 0, mutator_cpu_ns = 0,
                 mutator_elapsed_ns = 0, gc_cpu_ns = 0, gc_elapsed_ns = 0,
                 cpu_ns = 0, elapsed_ns = 0, gc = emptyGCDetails}
    |
479 | emptyRTSStats = Stats.RTSStats
    |                 ^^^^^^^^^^^^^^...
```

--- a/.travis.yml
+++ b/.travis.yml
@@ -28,24 +28,27 @@ before_cache:
 
 matrix:
   include:
-    - compiler: "ghc-8.4.3"
+    - compiler: "ghc-8.6.2"
     # env: TEST=--disable-tests BENCH=--disable-benchmarks
-      addons: {apt: {packages: [ghc-ppa-tools,cabal-install-2.2,ghc-8.4.3], sources: [hvr-ghc]}}
+      addons: {apt: {packages: [ghc-ppa-tools,cabal-install-2.4,ghc-8.6.2], sources: [hvr-ghc]}}
+    - compiler: "ghc-8.4.4"
+    # env: TEST=--disable-tests BENCH=--disable-benchmarks
+      addons: {apt: {packages: [ghc-ppa-tools,cabal-install-2.4,ghc-8.4.4], sources: [hvr-ghc]}}
     - compiler: "ghc-8.2.2"
     # env: TEST=--disable-tests BENCH=--disable-benchmarks
-      addons: {apt: {packages: [ghc-ppa-tools,cabal-install-2.2,ghc-8.2.2], sources: [hvr-ghc]}}
+      addons: {apt: {packages: [ghc-ppa-tools,cabal-install-2.4,ghc-8.2.2], sources: [hvr-ghc]}}
     - compiler: "ghc-8.0.2"
     # env: TEST=--disable-tests BENCH=--disable-benchmarks
-      addons: {apt: {packages: [ghc-ppa-tools,cabal-install-2.2,ghc-8.0.2], sources: [hvr-ghc]}}
+      addons: {apt: {packages: [ghc-ppa-tools,cabal-install-2.4,ghc-8.0.2], sources: [hvr-ghc]}}
     - compiler: "ghc-7.10.3"
     # env: TEST=--disable-tests BENCH=--disable-benchmarks
-      addons: {apt: {packages: [ghc-ppa-tools,cabal-install-2.2,ghc-7.10.3], sources: [hvr-ghc]}}
+      addons: {apt: {packages: [ghc-ppa-tools,cabal-install-2.4,ghc-7.10.3], sources: [hvr-ghc]}}
     - compiler: "ghc-7.8.4"
     # env: TEST=--disable-tests BENCH=--disable-benchmarks
-      addons: {apt: {packages: [ghc-ppa-tools,cabal-install-2.2,ghc-7.8.4], sources: [hvr-ghc]}}
+      addons: {apt: {packages: [ghc-ppa-tools,cabal-install-2.4,ghc-7.8.4], sources: [hvr-ghc]}}
     - compiler: "ghc-7.6.3"
     # env: TEST=--disable-tests BENCH=--disable-benchmarks
-      addons: {apt: {packages: [ghc-ppa-tools,cabal-install-2.2,ghc-7.6.3], sources: [hvr-ghc]}}
+      addons: {apt: {packages: [ghc-ppa-tools,cabal-install-2.4,ghc-7.6.3], sources: [hvr-ghc]}}
 
 before_install:
   - HC=${CC}
@@ -88,8 +91,8 @@ install:
 # any command which exits with a non-zero exit code causes the build to fail.
 script:
   # test that source-distributions can be generated
-  - (cd "." && cabal sdist)
-  - mv "."/dist/ekg-core-*.tar.gz ${DISTDIR}/
+  - cabal new-sdist all
+  - mv dist-newstyle/sdist/*.tar.gz ${DISTDIR}/
   - cd ${DISTDIR} || false
   - find . -maxdepth 1 -name '*.tar.gz' -exec tar -xvf '{}' \;
   - "printf 'packages: ekg-core-*/*.cabal\\n' > cabal.project"
@@ -107,7 +110,6 @@ script:
   - (cd ekg-core-* && cabal check)
 
   # haddock
-  - rm -rf ./dist-newstyle
   - if $HADDOCK; then cabal new-haddock -w ${HC} ${TEST} ${BENCH} all; else echo "Skipping haddock generation";fi
 
   # Build without installed constraints for packages in global-db

--- a/System/Metrics.hs
+++ b/System/Metrics.hs
@@ -370,6 +370,12 @@ sToMs s = round (s * 1000.0)
 --
 -- [@rts.gc.bytes_copied@] Number of bytes copied during GC
 --
+-- [@rts.gc.init_cpu_ms@] CPU time used by the init phase, in
+-- milliseconds. GHC 8.6+ only.
+--
+-- [@rts.gc.init_wall_ms@] Wall clock time spent running the init
+-- phase, in milliseconds. GHC 8.6+ only.
+--
 -- [@rts.gc.mutator_cpu_ms@] CPU time spent running mutator threads,
 -- in milliseconds. This does not include any profiling overhead or
 -- initialization.
@@ -423,6 +429,10 @@ registerGcMetrics store =
      , ("rts.gc.num_bytes_usage_samples"  , Counter . fromIntegral . Stats.major_gcs)
      , ("rts.gc.cumulative_bytes_used"    , Counter . fromIntegral . Stats.cumulative_live_bytes)
      , ("rts.gc.bytes_copied"             , Counter . fromIntegral . Stats.copied_bytes)
+#if MIN_VERSION_base(4,12,0)
+     , ("rts.gc.init_cpu_ms"              , Counter . nsToMs . Stats.init_cpu_ns)
+     , ("rts.gc.init_wall_ms"             , Counter . nsToMs . Stats.init_elapsed_ns)
+#endif
      , ("rts.gc.mutator_cpu_ms"           , Counter . nsToMs . Stats.mutator_cpu_ns)
      , ("rts.gc.mutator_wall_ms"          , Counter . nsToMs . Stats.mutator_elapsed_ns)
      , ("rts.gc.gc_cpu_ms"                , Counter . nsToMs . Stats.gc_cpu_ns)
@@ -491,6 +501,10 @@ emptyRTSStats = Stats.RTSStats
     , cumulative_par_max_copied_bytes      = 0
 # if MIN_VERSION_base(4,11,0)
     , cumulative_par_balanced_copied_bytes = 0
+# if MIN_VERSION_base(4,12,0)
+    , init_cpu_ns                          = 0
+    , init_elapsed_ns                      = 0
+# endif
 # endif
     , mutator_cpu_ns                       = 0
     , mutator_elapsed_ns                   = 0

--- a/ekg-core.cabal
+++ b/ekg-core.cabal
@@ -14,8 +14,9 @@ category:            System
 build-type:          Simple
 extra-source-files:  CHANGES.md
 cabal-version:       >=1.10
-tested-with:         GHC == 8.4.3,  GHC == 8.2.2, GHC == 8.0.2,
-                     GHC == 7.10.3, GHC == 7.8.4, GHC == 7.6.3
+tested-with:         GHC == 8.6.2, GHC == 8.4.4,  GHC == 8.2.2,
+                     GHC == 8.0.2, GHC == 7.10.3, GHC == 7.8.4,
+                     GHC == 7.6.3
 
 library
   exposed-modules:
@@ -33,7 +34,7 @@ library
 
   build-depends:
     ghc-prim < 0.6,
-    base >= 4.6 && < 4.12,
+    base >= 4.6 && < 4.13,
     containers >= 0.5 && < 0.7,
     text < 1.3,
     unordered-containers < 0.3

